src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,216 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size = 0;
+    private Node<T> first;
+    private Node<T> last;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            first = new Node<>(null, value, null);
+            last = first;
+        } else {
+            last.next = new Node<>(last, value, null);
+            last = last.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            checkIndexValidity(index);
+        }
+        if (index == 0 && size > 0) {
+            addToTheBeginning(value);
+        }
+        if (index == size) {
+            addToTheEnd(value);
+        } else if (index > 0) {
+            addToTheMiddle(value, index);
+        }
+    }
+
+    private void addToTheMiddle(T value, int index) {
+        if (index < size / 2) {
+            int i = 0;
+            for (MyLinkedList.Node<T> x = first; x != null; x = x.next) {
+                if (i == index) {
+                    createNodeAndItsLinks(value, x);
+                    break;
+                }
+                i++;
+            }
+        } else {
+            int i = size - 1;
+            for (MyLinkedList.Node<T> x = last; x != null; x = x.prev) {
+                if (i == index) {
+                    createNodeAndItsLinks(value, x);
+                    break;
+                }
+                i--;
+            }
+        }
+        size++;
+    }
+
+    private void createNodeAndItsLinks(T value, Node<T> x) {
+        Node<T> newNode = new Node<>(x.prev, value, x);
```

Comment:
You need to consider the case where `x.prev` is `null`. In this case, `x.prev.next` will cause a NullPointerException.